### PR TITLE
Subsidy calculation

### PR DESCRIPTION
**Pull request by chaosma**

### CODE REVIEW

File: core/ts/MaciState.ts [hunk 1]
```diff
@@ -717,6 +728,169 @@ class Poll {
...
+        const sbCommitment = hash3([stateRoot, ballotRoot, sbSalt])
+
+        const currentSubsidy = this.subsidy.map((x) => BigInt(x.toString()))
+        let currentSubsidyCommitment  = BigInt(0)
+        let currentSubsidySalt = BigInt(0)
+        let saltIndex = this.previousSubsidyIndexToString()
+        console.log(`prevIdx=${saltIndex}, curIdx=${this.rbi}-${this.cbi}`)
+        if (this.rbi !== 0 || this.cbi !== 0) {
+            currentSubsidySalt = BigInt(this.subsidySalts[saltIndex])
+            currentSubsidyCommitment = BigInt(genTallyResultCommitment(this.subsidy, currentSubsidySalt, this.treeDepths.voteOptionTreeDepth))
```
Comment (gurrpi) [inline core/ts/MaciState.ts:753]:
```suggestion
            currentSubsidyCommitment = BigInt(genTallyResultCommitment(this.subsidy, currentSubsidySalt, this.treeDepths.voteOptionTreeDepth).valueOf())
```
[The valueOf() method returns the wrapped primitive value of a BigInt object.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/valueOf)

File: core/ts/MaciState.ts [hunk 2]
```diff
@@ -717,6 +728,169 @@ class Poll {
...
+            rbi --
+            cbi = numBatches - 1
+        }
+        return rbi.toString() + "-" + cbi.toString()
+    }
+
+    public coefficientCalculation = (rowBallot: Ballot, colBallot: Ballot): BigInt  => {
+        let sum = BigInt(0)
+        for (let p = 0; p < this.maxValues.maxVoteOptions; p++) {
+            sum += BigInt(rowBallot.votes[p]) * BigInt(colBallot.votes[p])
```
Comment (gurrpi) [inline core/ts/MaciState.ts:841]:
```suggestion
            sum += BigInt(rowBallot.votes[p].valueOf()) * BigInt(colBallot.votes[p].valueOf())
```
[The valueOf() method returns the wrapped primitive value of a BigInt object.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/valueOf)

File: core/ts/MaciState.ts [hunk 3]
```diff
@@ -717,6 +728,169 @@ class Poll {
...
+        for (let i = 0; i < batchSize; i++) {
+            for (let j = 0; j < batchSize; j++) {
+                const row = rowStartIndex + i
+                const col = colStartIndex + j
+                const rowBallot = (row < this.ballots.length)?this.ballots[row]:emptyBallot
+                const colBallot = (col < this.ballots.length)?this.ballots[col]:emptyBallot
+
+                const kij = this.coefficientCalculation(rowBallot, colBallot)
+                for (let p = 0; p < this.maxValues.maxVoteOptions; p++) {
+                    const vip = BigInt(rowBallot.votes[p])
```
Comment (gurrpi) [inline core/ts/MaciState.ts:872]:
```suggestion
                    const vip = BigInt(rowBallot.votes[p].valueOf())
```
[The valueOf() method returns the wrapped primitive value of a BigInt object.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/valueOf)

File: core/ts/MaciState.ts [hunk 4]
```diff
@@ -717,6 +728,169 @@ class Poll {
...
+                    const vjp = BigInt(colBallot.votes[p])
```
Comment (gurrpi) [inline core/ts/MaciState.ts:873]:
```suggestion
                    const vjp = BigInt(colBallot.votes[p].valueOf())
```

File: core/ts/MaciState.ts [hunk 5]
```diff
@@ -717,6 +728,169 @@ class Poll {
...
+                    this.subsidy[p] = BigInt(this.subsidy[p]) + BigInt(kij) * vip * vjp
```
Comment (gurrpi) [inline core/ts/MaciState.ts:874]:
```suggestion
                    this.subsidy[p] = BigInt(this.subsidy[p].valueOf()) + BigInt(kij.valueOf()) * vip * vjp
```
[The valueOf() method returns the wrapped primitive value of a BigInt object.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/valueOf)

File: core/ts/MaciState.ts [hunk 6]
```diff
@@ -717,6 +728,169 @@ class Poll {
...
+            }
+        }
+
+        if (rowStartIndex === colStartIndex) {
+            for (let i = 0; i < batchSize; i++) {
+                const idx = rowStartIndex + i
+                const diagBallot = (idx < this.ballots.length)?this.ballots[idx]:emptyBallot
+                const kii = this.coefficientCalculation(diagBallot, diagBallot)
+                for (let p = 0; p < this.maxValues.maxVoteOptions; p++) {
+                    const vip = BigInt(diagBallot.votes[p])
```
Comment (gurrpi) [inline core/ts/MaciState.ts:885]:
```suggestion
                    const vip = BigInt(diagBallot.votes[p].valueOf())
```
[The valueOf() method returns the wrapped primitive value of a BigInt object.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/valueOf)

File: core/ts/MaciState.ts [hunk 7]
```diff
@@ -717,6 +728,169 @@ class Poll {
...
+                    this.subsidy[p] = BigInt(this.subsidy[p]) - BigInt(kii) * vip * vip
```
Comment (gurrpi) [inline core/ts/MaciState.ts:886]:
```suggestion
                    this.subsidy[p] = BigInt(this.subsidy[p].valueOf()) - BigInt(kii.valueOf()) * vip * vip
```
[The valueOf() method returns the wrapped primitive value of a BigInt object.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/valueOf)

**Review by gurrpi — COMMENTED** [review]
Figured out unwrapping `BigInt` to `bigint` fixes the [current build failure](https://github.com/appliedzkp/maci/pull/406/checks) of `maci-core`:
```
error TS2345: Argument of type 'BigInt' is not assignable to parameter of type 'string | number | bigint | boolean'.
```

**Comment by gurrpi** [comment]
Contract tests fails: 

```
 FAIL  ts/__tests__/MACI.test.ts (7.602 s)
  MACI
    Deployment
      ✓ MACI.stateTreeDepth should be correct (4 ms)
    Signups
      ✓ should sign up users (482 ms)
      ✓ signUp() shold fail when given an invalid pubkey (26 ms)
    Merging sign-ups should fail because of onlyPoll
      ✓ coordinator should not be able to merge the signUp AccQueue (9 ms)
    Deploy a Poll
      ✓ should set VKs and deploy a poll (229 ms)
      ✓ should set correct storage values (29 ms)
    Publish messages (vote + key-change)
      ✕ should publish a message to the Poll contract (1554 ms)
      ✓ shold not publish a message after the voting period (108 ms)
    Merge messages
      ✓ should revert if subtrees are not merged for StateAq (17 ms)
      ✓ coordinator should be able to merge the message AccQueue (751 ms)
      ✓ the message root must be correct (11 ms)
    Tally votes (negative test)
      ✓ tallyVotes() should fail as the messages have not been processed yet (17 ms)
    Process messages (negative test)
      ✓ processMessages() should fail if the state AQ has not been merged (15 ms)
    Merge sign-ups as the Poll
      ✓ The Poll should be able to merge the signUp AccQueue (1840 ms)
      ✓ the state root must be correct (3 ms)
    Process messages
      ✕ genProcessMessagesPackedVals() should generate the correct value (9 ms)
      ✕ processMessages() should update the state and ballot root commitment (4 ms)
    Tally votes
      ✓ genTallyVotesPackedVals() should generate the correct value (4 ms)
      ✕ tallyVotes() should update the tally commitment (240 ms)
    Generate MaciState from contract
      ✓ Should regenerate MaciState from on-chain information (211 ms)

  ● MACI › Publish messages (vote + key-change) › should publish a message to the Poll contract

    TypeError: Cannot read properties of undefined (reading 'rawPrivKey')

      377 |             expect(receipt.status).toEqual(1)
      378 |
    > 379 |             maciState.polls[pollId].publishMessage(message, keypair.pubKey)
          |                                     ^
      380 |         })
      381 |
      382 |         it('shold not publish a message after the voting period', async () => {

      at Function.Object.<anonymous>.Keypair.genEcdhSharedKey (node_modules/maci-domainobjs/ts/index.ts:344:41)
      at Poll.publishMessage (node_modules/maci-core/ts/MaciState.ts:200:35)
      at ts/__tests__/MACI.test.ts:379:37
      at step (ts/__tests__/MACI.test.ts:33:23)
      at Object.next (ts/__tests__/MACI.test.ts:14:53)
      at fulfilled (ts/__tests__/MACI.test.ts:5:58)
          at runMicrotasks (<anonymous>)

  ● MACI › Process messages › genProcessMessagesPackedVals() should generate the correct value

    TypeError: Cannot read properties of undefined (reading 'stateIndex')

      556 |
      557 |             poll = maciState.polls[pollId]
    > 558 |             generatedInputs = poll.processMessages(pollId)
          |                                    ^
      559 |         })
      560 |
      561 |         it('genProcessMessagesPackedVals() should generate the correct value', async () => {

      at Poll.genProcessMessagesCircuitInputsPartial (node_modules/maci-core/ts/MaciState.ts:499:51)
      at Poll.processMessages (node_modules/maci-core/ts/MaciState.ts:316:18)
      at ts/__tests__/MACI.test.ts:558:36
      at step (ts/__tests__/MACI.test.ts:33:23)
      at Object.next (ts/__tests__/MACI.test.ts:14:53)
      at fulfilled (ts/__tests__/MACI.test.ts:5:58)

  ● MACI › Process messages › processMessages() should update the state and ballot root commitment

    TypeError: Cannot read properties of undefined (reading 'stateIndex')

      556 |
      557 |             poll = maciState.polls[pollId]
    > 558 |             generatedInputs = poll.processMessages(pollId)
          |                                    ^
      559 |         })
      560 |
      561 |         it('genProcessMessagesPackedVals() should generate the correct value', async () => {

      at Poll.genProcessMessagesCircuitInputsPartial (node_modules/maci-core/ts/MaciState.ts:499:51)
      at Poll.processMessages (node_modules/maci-core/ts/MaciState.ts:316:18)
      at ts/__tests__/MACI.test.ts:558:36
      at step (ts/__tests__/MACI.test.ts:33:23)
      at Object.next (ts/__tests__/MACI.test.ts:14:53)
      at fulfilled (ts/__tests__/MACI.test.ts:5:58)

  ● MACI › Process messages › processMessages() should update the state and ballot root commitment

    TypeError: Cannot read properties of undefined (reading 'newSbCommitment')

      582 |             const tx = await pptContract.processMessages(
      583 |                 pollContractAddress,
    > 584 |                 generatedInputs.newSbCommitment,
          |                                 ^
      585 |                 [0, 0, 0, 0, 0, 0, 0, 0],
      586 |             )
      587 |

      at ts/__tests__/MACI.test.ts:584:33
      at step (ts/__tests__/MACI.test.ts:33:23)
      at Object.next (ts/__tests__/MACI.test.ts:14:53)
      at fulfilled (ts/__tests__/MACI.test.ts:5:58)

  ● MACI › Tally votes › tallyVotes() should update the tally commitment

    VM Exception while processing transaction: reverted with reason string 'PptE07'
        at PollProcessorAndTallyer.tallyVotes (contracts/Poll.sol:810)

      at HardhatNode._mineBlockWithPendingTxs (node_modules/hardhat/src/internal/hardhat-network/provider/node.ts:1650:23)
      at HardhatNode.mineBlock (node_modules/hardhat/src/internal/hardhat-network/provider/node.ts:459:16)
      at EthModule._sendTransactionAndReturnHash (node_modules/hardhat/src/internal/hardhat-network/provider/modules/eth.ts:1496:18)
      at HardhatNetworkProvider.request (node_modules/hardhat/src/internal/hardhat-network/provider/provider.ts:117:18)
      at EthersProviderWrapper.send (node_modules/@nomiclabs/hardhat-ethers/src/internal/ethers-provider-wrapper.ts:13:20)

  ● MACI › Tally votes › tallyVotes() should update the tally commitment

    expect.assertions(3)

    Expected three assertions to be called but received zero assertion calls.

      626 |
      627 |         it('tallyVotes() should update the tally commitment', async () => {
    > 628 |             expect.assertions(3)
          |                    ^
      629 |             const poll = maciState.polls[pollId]
      630 |             const generatedInputs = poll.tallyVotes(pollId)
      631 |

      at ts/__tests__/MACI.test.ts:628:20
      at step (ts/__tests__/MACI.test.ts:33:23)
      at Object.next (ts/__tests__/MACI.test.ts:14:53)
      at ts/__tests__/MACI.test.ts:8:71
      at Object.<anonymous>.__awaiter (ts/__tests__/MACI.test.ts:4:12)
      at Object.<anonymous> (ts/__tests__/MACI.test.ts:627:63)
```

File: cli/ts/setVerifyingKeys.ts [hunk 8]
```diff
@@ -95,6 +95,16 @@ const configureSubparser = (subparsers: any) => {
             help: 'The .zkey file for the vote tallying circuit. '
         }
     )
+    
+    createParser.addArgument(
+        ['-ss', '--subsidy-zkey'],
+        {
+            action: 'store',
+            type: 'string',
+            required: true,
```
Comment (gurrpi) [inline cli/ts/setVerifyingKeys.ts:104]: It causes [`cli/test1.sh`](https://github.com/appliedzkp/maci/blob/v1/cli/test1.sh) [fail](https://github.com/gurrpi/maci/runs/6047070045?check_suite_focus=true). As long as I remember from last weekly call, subsidy functionality should be optional.

**Comment by chnejohnson** [comment]
> It causes [`cli/test1.sh`](https://github.com/appliedzkp/maci/blob/v1/cli/test1.sh) [fail](https://github.com/gurrpi/maci/runs/6047070045?check_suite_focus=true). As long as I remember from last weekly call, subsidy functionality should be optional.

Agree, this feature had better be optional. People who run QV don't need to do subsidy calculation.

**Comment by chaosma** [comment]
> Figured out unwrapping `BigInt` to `bigint` fixes the [current build failure](https://github.com/appliedzkp/maci/pull/406/checks) of `maci-core`:
> 
> ```
> error TS2345: Argument of type 'BigInt' is not assignable to parameter of type 'string | number | bigint | boolean'.
> ```

Thanks! updated these changes

File: cli/ts/setVerifyingKeys.ts (hunk 8 above)
Comment (chaosma) [reply cli/ts/setVerifyingKeys.ts]: I made the subsidy calculation as optional now. Could you help to add test case for subsidy calculation as well? test.sh is the script to test multiple batches including subsidy calculation. I also updated the outdated test1.sh to test without subsidy case.

**Comment by chaosma** [comment]
> > It causes [`cli/test1.sh`](https://github.com/appliedzkp/maci/blob/v1/cli/test1.sh) [fail](https://github.com/gurrpi/maci/runs/6047070045?check_suite_focus=true). As long as I remember from last weekly call, subsidy functionality should be optional.
> 
> Agree, this feature had better be optional. People who run QV don't need to do subsidy calculation.

Yeah, I have made the subsidy calculation optional.

File: cli/ts/setVerifyingKeys.ts (hunk 8 above)
Comment (gurrpi) [reply cli/ts/setVerifyingKeys.ts]: Sure, filed an issue: https://github.com/appliedzkp/maci/issues/413

**Pull request merged**